Source/WebKit/Platform/IPC/glib/ConnectionGLib.cpp

```diff
@@ -0,0 +1,575 @@
+
+    m_readSocketMonitor.start(m_socket.get(), G_IO_IN, m_connectionQueue->runLoop(), m_cancellable.get(), [protectedThis] (GIOCondition condition) -> gboolean {
+        if (condition & G_IO_HUP || condition & G_IO_ERR || condition & G_IO_NVAL) {
+            protectedThis->connectionDidClose();
+            return G_SOURCE_REMOVE;
+        }
+
+        if (condition & G_IO_IN) {
+            protectedThis->readyReadHandler();
+            return G_SOURCE_CONTINUE;
+        }
+
+        ASSERT_NOT_REACHED();
+        return G_SOURCE_REMOVE;
+    });
+
+    // Schedule a call to readyReadHandler. Data may have arrived before installation of the signal handler.
+    m_connectionQueue->dispatch([protectedThis] {
+        protectedThis->readyReadHandler();
+    });
+}
+
+bool Connection::platformCanSendOutgoingMessages() const
+{
+#if OS(ANDROID)
+    return !m_hasPendingOutputMessage && m_outgoingHardwareBuffers.isEmpty();
+#else
+    return !m_hasPendingOutputMessage;
+#endif
+}
+
+bool Connection::sendOutgoingMessage(UniqueRef<Encoder>&& encoder)
+{
+    static_assert(sizeof(MessageInfo) + s_attachmentMaxAmount * sizeof(size_t) <= s_messageMaxSize, "Attachments fit to message inline");
+
+    UnixMessage outputMessage(encoder.get());
+    if (outputMessage.attachments().size() > (s_attachmentMaxAmount - 1)) {
+        ASSERT_NOT_REACHED();
+        return false;
+    }
+
+    size_t messageSizeWithBodyInline = sizeof(MessageInfo) + (outputMessage.attachments().size() * sizeof(AttachmentInfo)) + outputMessage.bodySize();
+    if (messageSizeWithBodyInline > s_messageMaxSize && outputMessage.bodySize()) {
+        if (!outputMessage.setBodyOutOfLine())
+            return false;
+    }
+
+    return sendOutputMessage(WTFMove(outputMessage));
+}
+
+bool Connection::sendOutputMessage(UnixMessage&& outputMessage)
+{
+#if OS(ANDROID)
+    RELEASE_ASSERT(m_outgoingHardwareBuffers.isEmpty());
+    Vector<RefPtr<AHardwareBuffer>, 2> hardwareBuffers;
+#endif
+    ASSERT(!m_hasPendingOutputMessage);
+
+    auto& messageInfo = outputMessage.messageInfo();
+    const auto& attachments = outputMessage.attachments();
+    GOutputVector outputVector[3];
+    int outputVectorLength = 0;
+
+    outputVector[outputVectorLength++] = { reinterpret_cast<void*>(&messageInfo), sizeof(messageInfo) };
+    GRefPtr<GSocketControlMessage> controlMessage;
+    if (!attachments.isEmpty()) {
+        Vector<AttachmentInfo> attachmentInfo(attachments.size());
+        Vector<int> fds;
+        fds.reserveInitialCapacity(attachments.size());
+        for (size_t i = 0; i < attachments.size(); ++i) {
+#if OS(ANDROID)
+            RELEASE_ASSERT(attachmentInfo[i].type() == AttachmentInfo::Type::Unset);
+            switchOn(attachments[i],
+                [&](const UnixFileDescriptor& fd) {
+                    attachmentInfo[i].setType(AttachmentInfo::Type::FileDescriptor);
+                    if (fd)
+                        fds.append(attachments[i].value());
```

Comment:
This should use `fd`:

```suggestion
                        fds.append(fd.value());
```